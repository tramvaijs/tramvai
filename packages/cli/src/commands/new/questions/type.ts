import any from '@tinkoff/utils/array/any';

type Type = 'app' | 'child-app';

const choices = [
  {
    name: 'Standalone tramvai app',
    value: 'app',
  },
  {
    name: 'Microfrontend implemented by tramvai child-app',
    value: 'child-app',
  },
];

const typeQuestion = (answer: string) => ({
  type: 'list' as const,
  name: 'type' as const,
  message: 'Choose a type for your project',
  choices,
  when: () => !any<{ name: string; value: string }>((choice) => choice.value === answer)(choices),
  default: 'app' as const,
});

export { Type, typeQuestion };
