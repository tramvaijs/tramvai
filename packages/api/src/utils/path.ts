import path from 'node:path';

export const resolveAbsolutePathForFile = ({
  file,
  sourceDir,
  rootDir,
}: {
  file: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(file)) {
    return file;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, file);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, file);
  }
  return path.resolve(rootDir, sourceDir, file);
};

export const resolveAbsolutePathForFolder = ({
  folder,
  rootDir,
}: {
  folder: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(folder)) {
    return folder;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, folder);
  }
  return path.resolve(rootDir, folder);
};
