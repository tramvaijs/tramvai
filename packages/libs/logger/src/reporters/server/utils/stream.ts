export function writeStream(data, stream) {
  const write = stream.__write || stream.write;

  return write.call(stream, data);
}
