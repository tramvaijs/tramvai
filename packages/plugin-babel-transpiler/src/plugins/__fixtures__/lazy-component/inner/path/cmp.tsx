import React from 'react';

const InnerCmp = () => {
  return <div>inner cmp</div>;
}

export default InnerCmp;
