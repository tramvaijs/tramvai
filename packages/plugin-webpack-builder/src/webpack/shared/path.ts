import path from 'node:path';

export const resolveAbsolutePathForEntry = ({
  entry,
  sourceDir,
  rootDir,
}: {
  entry: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(entry)) {
    return entry;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, entry);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, entry);
  }
  return path.resolve(rootDir, sourceDir, entry);
};

export const resolveAbsolutePathForOutput = ({
  output,
  rootDir,
}: {
  output: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(output)) {
    return output;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, output);
  }
  return path.resolve(rootDir, output);
};
