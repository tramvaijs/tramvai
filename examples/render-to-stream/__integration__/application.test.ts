import type { StartCliResult } from '@tramvai/test-integration';
import { sleep, startCli } from '@tramvai/test-integration';
import { initPlaywright } from '@tramvai/test-pw';
import path from 'path';
import { Writable } from 'stream';

jest.setTimeout(30000);

class WritableBuffer extends Writable {
  private chunks: string[] = [];

  getChunks() {
    return this.chunks;
  }

  _write(chunk, encoding, callback) {
    this.chunks.push(chunk.toString('utf-8'));
    callback();
  }
}

describe('render-to-stream', () => {
  let app: StartCliResult;

  beforeAll(async () => {
    app = await startCli('render-to-stream', {
      rootDir: path.resolve(__dirname, '../'),
    });
  }, 80000);

  afterAll(() => {
    return app.close();
  });

  describe('streaming response', () => {
    // eslint-disable-next-line jest/no-done-callback
    it('deferred page', (done) => {
      const stream = new WritableBuffer();

      const initialChunksContent = [
        '<!DOCTYPE html>',
        // actions, resolved before response
        `window.__TRAMVAI_DEFERRED_ACTIONS['fastDeferred'].resolve({"data":"ok"});</script>`,
        `window.__TRAMVAI_DEFERRED_ACTIONS['failedFastDeferred'].reject({"message":"Failed Fast Deferred"`,
        '<div class="application">',
        // lazy component outside Suspense
        'Deferred Page',
        `<script>typeof window.__TRAMVAI_DEFERRED_HYDRATION === 'function' ? window.__TRAMVAI_DEFERRED_HYDRATION() : window.__TRAMVAI_DEFERRED_HYDRATION = true;</script>`,
      ];
      const deferredChunksContent = [
        // deferred actions promises teleportation
        `<script>window.__TRAMVAI_DEFERRED_ACTIONS['longDeferred'].resolve({"data":"ok"});</script>`,
        `<script>window.__TRAMVAI_DEFERRED_ACTIONS['failedDeferred'].reject({"message":"Failed Deferred"`,
        // lazy component inside Suspense, blocking assets
        'components-features-Data-Data.chunk.css',
        'components-features-Data-Data.chunk.js',
        `Response: ok`,
        '</body></html>',
      ];

      setTimeout(() => {
        const initialHTML = stream.getChunks().join('\n');

        // all critical scripts async for streaming
        expect(initialHTML.match(/data-critical="true" async="async"/g)?.length).toBe(5);

        initialChunksContent.forEach((chunk) => {
          expect(initialHTML).toContain(chunk);
        });

        deferredChunksContent.forEach((chunk) => {
          expect(initialHTML).not.toContain(chunk);
        });
      }, 500);

      app
        .request('/deferred/')
        .expect(200)
        .expect('Transfer-Encoding', 'chunked')
        .pipe(stream)
        .on('finish', (error) => {
          if (error) {
            return done(error);
          }

          const finalHTML = stream.getChunks().join('\n');

          deferredChunksContent.forEach((chunk) => {
            expect(finalHTML).toContain(chunk);
          });

          expect.assertions(19);

          done();
        });
    });
  });

  describe('server-side rendering', () => {
    it('main page', async () => {
      const { parsed } = await app.render('/');

      expect(parsed.innerHTML).toMatchSnapshot();
    });

    it('deferred page', async () => {
      const { parsed } = await app.render('/deferred/');

      expect(parsed.innerHTML).toMatchSnapshot();
    });

    it('non-deferred page', async () => {
      const { parsed } = await app.render('/non-deferred/');

      expect(parsed.innerHTML).toMatchSnapshot();
    });
  });

  describe('client hydration', () => {
    it('main page', async () => {
      const { browser, getPageWrapper } = await initPlaywright(app.serverUrl);

      const { page } = await getPageWrapper(app.serverUrl);

      expect(await page.$eval('.application', (node) => (node as HTMLElement).innerText))
        .toMatchInlineSnapshot(`
        "Tramvai ðŸ¥³
        Main
        Second
        Deferred
        Deferred State
        Non-deferred
        Deferred Foo
        Deferred Bar
        Deferred Baz
        Main Page
        Child Component
        Error boundary
        this Footer in render-to-stream

        This is modal for index page!"
      `);

      await browser.close();
    });

    it('deferred page', async () => {
      const { browser, getPageWrapper } = await initPlaywright(app.serverUrl);

      const { page } = await getPageWrapper(`${app.serverUrl}/deferred/`);

      expect(await page.$eval('.application', (node) => (node as HTMLElement).innerText))
        .toMatchInlineSnapshot(`
        "Tramvai ðŸ¥³
        Main
        Second
        Deferred
        Deferred State
        Non-deferred
        Deferred Foo
        Deferred Bar
        Deferred Baz
        Deferred Page
        Response: ok
        Response: ok
        Error: Failed Fast Deferred
        Error: Failed Deferred
        Error: Deferred Action Abort
        this Footer in render-to-stream"
      `);

      await browser.close();
    });

    it('deferred page with parameter', async () => {
      const { browser, getPageWrapper } = await initPlaywright(app.serverUrl);

      const { page } = await getPageWrapper(`${app.serverUrl}/deferred/foo/`);

      expect(await page.$eval('.application', (node) => (node as HTMLElement).innerText))
        .toMatchInlineSnapshot(`
        "Tramvai ðŸ¥³
        Main
        Second
        Deferred
        Deferred State
        Non-deferred
        Deferred Foo
        Deferred Bar
        Deferred Baz
        Deferred Id Page
        Response: foo
        this Footer in render-to-stream"
      `);

      await browser.close();
    });

    it('SPA-transition between same deferred page with different parameters', async () => {
      const { browser, getPageWrapper } = await initPlaywright(app.serverUrl);

      const { page, router } = await getPageWrapper(`${app.serverUrl}/deferred/foo/`);

      await router.navigate('/deferred/bar/');

      expect(await page.$eval('.application', (node) => (node as HTMLElement).innerText))
        .toMatchInlineSnapshot(`
        "Tramvai ðŸ¥³
        Main
        Second
        Deferred
        Deferred State
        Non-deferred
        Deferred Foo
        Deferred Bar
        Deferred Baz
        Deferred Id Page
        Loading long...
        this Footer in render-to-stream"
      `);

      await sleep(1000);

      expect(await page.$eval('.application', (node) => (node as HTMLElement).innerText))
        .toMatchInlineSnapshot(`
        "Tramvai ðŸ¥³
        Main
        Second
        Deferred
        Deferred State
        Non-deferred
        Deferred Foo
        Deferred Bar
        Deferred Baz
        Deferred Id Page
        Response: bar
        this Footer in render-to-stream"
      `);

      await browser.close();
    });
  });
});
