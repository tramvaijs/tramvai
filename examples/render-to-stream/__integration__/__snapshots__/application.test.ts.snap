// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render-to-stream server-side rendering deferred page 1`] = `
"<!DOCTYPE html>
<html class="no-js" lang="ru">
  <head>
    <meta charset="UTF-8" >
    
    <title data-meta-dynamic="true">Tramvai</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    

    

    

    
    <link crossorigin="anonymous"
      data-critical="true"
      rel="stylesheet"
      href="\${STATIC_URL}/dist/client/components-features-Title-Title.chunk.css"
    >
    <link crossorigin="anonymous" data-critical="true" rel="stylesheet" href="\${STATIC_URL}/dist/client/platform.css" >
    

    

    

    
    <script id="__TRAMVAI_STATE__" type="application/json"></script>
    

    
    <script></script>
    <script></script>
    <script></script>
    <script></script>
    <script></script>
    <script></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/@_routes_deferred_index.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/components-features-Title-Title.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/hmr.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  </head>
  <body>
    

    

    
    <div class="application">
      <div>
        <h1>
          Tramvai
          <span role="img" aria-label="Salute">ðŸ¥³</span>
        </h1>
        <ul>
          <li><a href="/">Main</a></li>
          <li><a href="/second/">Second</a></li>
          <li><a href="/deferred/">Deferred</a></li>
          <li><a href="/deferred-state/">Deferred State</a></li>
          <li><a href="/non-deferred/">Non-deferred</a></li>
          <li><a href="/deferred/foo/">Deferred Foo</a></li>
          <li><a href="/deferred/bar/">Deferred Bar</a></li>
          <li><a href="/deferred/baz/">Deferred Baz</a></li>
        </ul>
      </div>
      <h3 class="Title__title_jjG8q">Deferred Page</h3>
      <template id="B:0"></template>
      <div>Loading long...</div>
      
      <div>
        Response:
        ok
      </div>
      
      <div>
        Error:
        Failed Fast Deferred
      </div>
      <template id="B:1"></template>
      <div>Loading long error...</div>
      <template id="B:2"></template>
      <div>Loading aborted error...</div>
      
      <div class="Footer__footer_L7U1b"><div>this Footer in render-to-stream</div></div>
      <script></script>
      <script></script>
      <script></script>
      <link rel="stylesheet"
        href="\${STATIC_URL}/dist/client/components-features-Data-Data.chunk.css"
        crossorigin="anonymous"
        data-critical="true"
      >
      <script src="\${STATIC_URL}/dist/client/components-features-Data-Data.chunk.js"
        charset="utf-8"
        crossorigin="anonymous"
        data-critical="true"
      ></script>
      <div hidden id="S:0"><div class="Data__data_Q_cJh">Response: ok</div></div>
      <script></script>
      <div hidden id="S:1">
        <div>
          Error:
          Failed Deferred
        </div>
      </div>
      <script></script>
      <script></script>
      <script></script>
    </div>
    

    

    

    

    

    

    
  </body>
</html>
"
`;

exports[`render-to-stream server-side rendering main page 1`] = `
"<!DOCTYPE html>
<html class="no-js" lang="ru">
  <head>
    <meta charset="UTF-8" >
    
    <title data-meta-dynamic="true">Tramvai</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    

    

    

    
    <link crossorigin="anonymous"
      data-critical="true"
      rel="stylesheet"
      href="\${STATIC_URL}/dist/client/@_routes_index.chunk.css"
    >
    <link crossorigin="anonymous" data-critical="true" rel="stylesheet" href="\${STATIC_URL}/dist/client/platform.css" >
    

    

    

    
    <script id="__TRAMVAI_STATE__" type="application/json"></script>
    

    
    <script></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/@_routes_index.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/hmr.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  </head>
  <body>
    

    

    
    <div class="application">
      <div>
        <h1>
          Tramvai
          <span role="img" aria-label="Salute">ðŸ¥³</span>
        </h1>
        <ul>
          <li><a href="/">Main</a></li>
          <li><a href="/second/">Second</a></li>
          <li><a href="/deferred/">Deferred</a></li>
          <li><a href="/deferred-state/">Deferred State</a></li>
          <li><a href="/non-deferred/">Non-deferred</a></li>
          <li><a href="/deferred/foo/">Deferred Foo</a></li>
          <li><a href="/deferred/bar/">Deferred Bar</a></li>
          <li><a href="/deferred/baz/">Deferred Baz</a></li>
        </ul>
      </div>
      <div>
        Main Page
        <div>Child Component</div>
        <template data-msg="Error"
          data-stck="
    at ErrorComponent (webpack://@tramvai/cli/../../examples/render-to-stream/src/routes/index.tsx?:[LINE]:[SYMBOL])
    at Suspense
    at ErrorBoundary (webpack://@tramvai/cli/../../examples/render-to-stream/src/routes/index.tsx?:[LINE]:[SYMBOL])
    at div
    at MainPage
    at InnerLoadable (webpack://@tramvai/cli/../../node_modules/@loadable/component/dist/loadable.esm.js?:[LINE]:[SYMBOL])
    at LoadableWithChunkExtractor
    at Loadable
    at UniversalErrorBoundary (webpack://@tramvai/cli/../tramvai/react/lib/error/UniversalErrorBoundary.js?:[LINE]:[SYMBOL])
    at PageErrorBoundary (webpack://@tramvai/cli/../modules/render/lib/react/pageErrorBoundary.js?:[LINE]:[SYMBOL])
    at PageRenderComponent (webpack://@tramvai/cli/../modules/render/lib/react/root.js?:[LINE]:[SYMBOL])
    at RenderChildrenComponent (webpack://@tramvai/cli/../modules/render/lib/shared/LayoutModule.js?:[LINE]:[SYMBOL])
    at NestedLayoutRenderComponent (webpack://@tramvai/cli/../modules/render/lib/react/root.js?:[LINE]:[SYMBOL])
    at RenderChildren (webpack://@tramvai/cli/../libs/tinkoff-layout/lib/layout.js?:[LINE]:[SYMBOL])
    at RenderChildren (webpack://@tramvai/cli/../libs/tinkoff-layout/lib/layout.js?:[LINE]:[SYMBOL])
    at RenderChildren (webpack://@tramvai/cli/../libs/tinkoff-layout/lib/layout.js?:[LINE]:[SYMBOL])
    at Layout (webpack://@tramvai/cli/../libs/tinkoff-layout/lib/layout.js?:[LINE]:[SYMBOL])
    at LayoutRenderComponent (webpack://@tramvai/cli/../modules/render/lib/react/root.js?:[LINE]:[SYMBOL])
    at Root (webpack://@tramvai/cli/../modules/render/lib/react/root.js?:[LINE]:[SYMBOL])
    at Provider (webpack://@tramvai/cli/../tramvai/state/lib/connect/Provider.js?:[LINE]:[SYMBOL])
    at DefaultProvider (webpack://@tramvai/cli/../libs/router/lib/components/react/providers/default.js?:[LINE]:[SYMBOL])
    at ChunkExtractorManager (webpack://@tramvai/cli/../../node_modules/@loadable/server/lib/ChunkExtractorManager.js?:[LINE]:[SYMBOL])"
        ></template>Error fallback
      </div>
      <div class="Footer__footer_L7U1b">
        <div>this Footer in render-to-stream</div>
        <p class="main__MainModal_wPdhA">This is modal for index page!</p>
      </div>
      <script></script>
    </div>
    

    

    

    

    

    

    
  </body>
</html>
"
`;

exports[`render-to-stream server-side rendering non-deferred page 1`] = `
"<!DOCTYPE html>
<html class="no-js" lang="ru">
  <head>
    <meta charset="UTF-8" >
    
    <title data-meta-dynamic="true">Tramvai</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    

    

    

    
    <link crossorigin="anonymous" data-critical="true" rel="stylesheet" href="\${STATIC_URL}/dist/client/platform.css" >
    

    

    

    
    <script id="__TRAMVAI_STATE__" type="application/json"></script>
    

    
    <script></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/@_routes_non-deferred_index.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/hmr.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  </head>
  <body>
    

    

    
    <div class="application">
      <div>
        <h1>
          Tramvai
          <span role="img" aria-label="Salute">ðŸ¥³</span>
        </h1>
        <ul>
          <li><a href="/">Main</a></li>
          <li><a href="/second/">Second</a></li>
          <li><a href="/deferred/">Deferred</a></li>
          <li><a href="/deferred-state/">Deferred State</a></li>
          <li><a href="/non-deferred/">Non-deferred</a></li>
          <li><a href="/deferred/foo/">Deferred Foo</a></li>
          <li><a href="/deferred/bar/">Deferred Bar</a></li>
          <li><a href="/deferred/baz/">Deferred Baz</a></li>
        </ul>
      </div>
      <div>Loading...</div>
      <div class="Footer__footer_L7U1b"><div>this Footer in render-to-stream</div></div>
      <script></script>
    </div>
    

    

    

    

    

    

    
  </body>
</html>
"
`;
